### DISCUSSION TITLE
Theming component 🎨

### DISCUSSION THREAD

**Discussion by yinonov** [Vonage/vivid#856] [New Developments]
Develop a wrapper component to apply theming to inner components
Apply css variables only to define theming context
note that any css styling props will [not apply to tags deeper than 1st level nesting ](https://stackoverflow.com/a/61631668/3027703)

this has been in plan in the past but was dropped due to the barriers mentioned above.

the updated plan is to only provide the theming exactly as scheme does but in a component modular practice
```html
<vwc-theme dark alternate>
  <form>
    <vwc-textfield></vwc-textfield>
    <vwc-button></vwc-button>
  </form>
</vwc-theme>
```

**Note**: [Fast components](https://www.fast.design/docs/components/getting-started#add-the-design-system-provider) (Microsoft's), [Spectrum](https://opensource.adobe.com/spectrum-web-components/components/theme) (Adobe) provide that same practice

This is a 1 stop shop for

* color scheme
* scale (sizing, typography, block-size, shaping)

**Comment by YonatanKra** [comment]
It is a missing feature and I think can be very useful.
Then again - do we have a use case for it?
If we go open source, we'd probably tackle this - but inside Vonage? Dunno...

**Reply by yinonov** [reply]
It will provide the same output as scheme. Only in an encapsulated modular fashion

**Comment by gullerya** [comment]
I'd put the following cases as a validational:
- will we allow / enforce more than one `vwc-theme` in the whole app? given that it is including `alternate`, probably yes, but can we / shall we ensure that `dark` is not mixed with `light` within it?
- even when not parent child case - will we allow sibling-like `vwc-theme` in the whole app? can we / shall we ensure that the theme is the same for all of them?
- how do we see integration flow from the consumer perspective: is it spread all of it around HTML and later on move them all up till the `body` is last wrapped?

**Reply by yinonov** [reply]
good point on light/dark, we dont want this to stir. initializing a theme in a session should be idempotent. 
as for integration, I'm saying ATM let's steer away from context as it feels like meddling in LOBs apps is too challenging while trying to hold the rope at its both ends. so yes, expand the region of wrapping gradually.

the vision is to later on provide more identities such

- `<vwc-typography>`
- `<vwc-spacing dense>`
- `<vwc-shape>`
- etc'...

**Comment by tveinfeld** [comment]
Interesting concept 👍 what is the advantage over using CSS classes defining variables?

**Reply by yinonov** [reply]
https://github.com/Vonage/vivid/discussions/856#discussioncomment-774224

**Reply by tveinfeld** [reply]
why is encapsulation important?

in the example above, would the css vars applied on `vwc-textfield` by `vwc-theme` pierce through its shadow dom? if so, how does the encapsulation quality manifests?

**Reply by yinonov** [reply]
[this](https://lit.dev/playground/#project=W3sibmFtZSI6Im15LXRoZW1lLnRzIiwiY29udGVudCI6ImltcG9ydCB7aHRtbCwgY3NzLCBMaXRFbGVtZW50fSBmcm9tICdsaXQnO1xuaW1wb3J0IHtjdXN0b21FbGVtZW50LCBwcm9wZXJ0eX0gZnJvbSAnbGl0L2RlY29yYXRvcnMuanMnO1xuXG5AY3VzdG9tRWxlbWVudCgnbXktdGhlbWUnKVxuZXhwb3J0IGNsYXNzIE15VGhlbWUgZXh0ZW5kcyBMaXRFbGVtZW50IHtcbiAgc3RhdGljIHN0eWxlcyA9IGNzc2A6aG9zdCB7IFxuICAgICAgLS1jb2xvci1wcmltYXJ5OiBsaWdodGJsdWU7XG4gICAgICAtLWNvbG9yLW9uLXByaW1hcnk6IGJsYWNrO1xuICAgICAgfWA7XG5cbiAgcmVuZGVyKCkge1xuICAgIHJldHVybiBodG1sYDxzbG90Pjwvc2xvdD5gO1xuICB9XG59XG4ifSx7Im5hbWUiOiJpbmRleC5odG1sIiwiY29udGVudCI6IjwhRE9DVFlQRSBodG1sPlxuPGhlYWQ-XG4gIDxzY3JpcHQgdHlwZT1cIm1vZHVsZVwiIHNyYz1cIi4vbXktdGhlbWUuanNcIj48L3NjcmlwdD5cbiAgPHNjcmlwdCB0eXBlPVwibW9kdWxlXCIgc3JjPVwiLi9teS1idXR0b24uanNcIj48L3NjcmlwdD5cbjwvaGVhZD5cbjxib2R5PlxuICA8bXktdGhlbWU-XG4gICAgPGZvcm0-XG4gICAgICA8bXktYnV0dG9uIGxhYmVsPVwiQ2xpY2sgbWVcIj48L215LWJ1dHRvbj5cbiAgICA8L2Zvcm0-XG4gIDwvbXktdGhlbWU-XG48L2JvZHk-XG4ifSx7Im5hbWUiOiJteS1idXR0b24udHMiLCJjb250ZW50IjoiaW1wb3J0IHtodG1sLCBjc3MsIExpdEVsZW1lbnR9IGZyb20gJ2xpdCc7XG5pbXBvcnQge2N1c3RvbUVsZW1lbnQsIHByb3BlcnR5fSBmcm9tICdsaXQvZGVjb3JhdG9ycy5qcyc7XG5cbkBjdXN0b21FbGVtZW50KCdteS1idXR0b24nKVxuZXhwb3J0IGNsYXNzIE15QnV0dG9uIGV4dGVuZHMgTGl0RWxlbWVudCB7XG4gIHN0YXRpYyBzdHlsZXMgPSBjc3NgXG4gICAgICBidXR0b24ge1xuICAgICAgYmFja2dyb3VuZC1jb2xvcjogdmFyKC0tY29sb3ItcHJpbWFyeSk7XG4gICAgICBjb2xvcjogdmFyKC0tY29sb3Itb24tcHJpbWFyeSk7XG4gICAgICBib3JkZXI6IHVuc2V0O1xuICAgICAgYmxvY2stc2l6ZTogNDBweFxuICAgICAgfWA7XG5cbiAgQHByb3BlcnR5KClcbiAgbGFiZWw_O1xuXG4gIHJlbmRlcigpIHtcbiAgICByZXR1cm4gaHRtbGA8YnV0dG9uPiR7dGhpcy5sYWJlbH08L2J1dHRvbj5gO1xuICB9XG59XG4ifV0) might help illustrate it.

I feel it makes a more declarative code in the html.

consumers would be able to easily grasp and implements UI regions. well, the plan is to provide only ***vvd*** prefixed css variables so encapsulation isn't a big issue and might not affect any style incidentally. but it will still keep a pristine DOM structure

**Comment by yinonov** [comment]
Introducing an idea to provide design system identities as components rather than esm that apply styles (css variables only) on the html scope.
This should be referred to as a concept on how to bind components styles to predefined variants of choice.

```html
<!-- theme ideally would be integrated only once in an app -->
<vwc-theme theme="light">
   <!-- typography ideally would be integrated only once in an app -->
   <vwc-typography>
      <vwc-sizing>
         <main>
            <vwc-top-app-bar>
               <!-- theme toggle button would set the closest vwc-theme ancestor (if exist) -->
               <vwc-theme-toggle-button slot="action"></vwc-theme-toggle-button>
            </vwc-top-app-bar>
            <h1>Welcome to our app</h1>
            <!-- notice this deeper scope and second usage of sizing to apply large size on button instead of the current button API 'enlarged' button as a low level component can drop support of dense / enlarged properties, rather would refer its block-size value to a css variable provided by `vwc - sizing` (it can obviously default to some hardcoded value of choice) -->
            <vwc-sizing size="large">
               <vwc-button>Click me</vwc-button>
            </vwc-sizing>
         </main>
         <aside>
            <vwc-alternate>
               <!-- aside content -->
            </vwc-alternate>
         </aside>
      </vwc-sizing>
   </vwc-typography>
</vwc-theme>
```

disadvantages:

- Extra html for consumers to handle
- Might result in breaking changes (unless able to precedence the theme component's specificity over affected components)

**Reply by yinonov** [reply]
### challenging use case
buttons can have various shapes by default
```html
<vwc-button>Click me</vwc-button>
<vwc-icon-button icon="profile"></vwc-icon-button>
```
The above default **button** is rounded in corners while **icon-button** is circled
how will `<vwc-shape>` affect the 2 if has a default of rounded? how can we apply 2 shapes in same context?
